src/main/kotlin/org/rust/cargo/runconfig/CargoTestRunState.kt

```diff
@@ -0,0 +1,81 @@
+import com.intellij.execution.testframework.autotest.ToggleAutoTestAction
+import com.intellij.execution.testframework.sm.SMTestRunnerConnectionUtil
+import com.intellij.execution.testframework.sm.runner.ui.SMTRunnerConsoleView
+import org.rust.cargo.runconfig.command.CargoCommandConfiguration
+import org.rust.cargo.runconfig.test.CargoTestConsoleProperties
+import com.intellij.openapi.util.Getter
+import org.jetbrains.annotations.TestOnly
+import org.rust.cargo.toolchain.CargoCommandLine
+
+class CargoTestRunState(
+    environment: ExecutionEnvironment,
+    config: CargoCommandConfiguration.CleanConfiguration.Ok
+) : CargoRunState(environment, config) {
+    override fun execute(executor: Executor, runner: ProgramRunner<*>): ExecutionResult {
+        val processHandler = startProcess()
+
+        val consoleProperties = CargoTestConsoleProperties(environment.runProfile as RunConfiguration, executor)
+        val consoleView = SMTestRunnerConnectionUtil.createAndAttachConsole(
+            "Cargo Test", processHandler, consoleProperties) as SMTRunnerConsoleView
+
+        createFilters().forEach { consoleView.addMessageFilter(it) }
+
+        val executionResult = DefaultExecutionResult(consoleView, processHandler)
+        val rerunFailedTestsAction = consoleProperties.createRerunFailedTestsAction(consoleView)
+        if (rerunFailedTestsAction != null) {
+            rerunFailedTestsAction.setModelProvider(Getter<TestFrameworkRunningModel> { consoleView.resultsViewer })
+            executionResult.setRestartActions(rerunFailedTestsAction, ToggleAutoTestAction())
+        } else {
+            executionResult.setRestartActions(ToggleAutoTestAction())
+        }
+        return executionResult
+    }
+
+    override fun prepareCommandLine(): CargoCommandLine =
+        commandLine.copy(additionalArguments = patchArgs(commandLine.additionalArguments))
+
+    companion object {
+        /**
+         * This method is @TestOnly because it should be private, but there are unit tests for it.
+         *
+         * Libtest likes to yell when there are duplicated
+         * args, so we cannot just insert our defaults :(
+         */
+        @TestOnly
+        fun patchArgs(args: List<String>): List<String> {
+            val ddIdx = args.indexOf("--")
+            val (cargoArgs, libtestArgs) =
+                if (ddIdx < 0) {
+                    args.toMutableList() to mutableListOf()
+                } else {
+                    args.subList(0, ddIdx).toMutableList() to args.subList(ddIdx + 1, args.size).toMutableList()
+                }
+
+            val noFormatJson = "--format=json"
+            val noUnstableOptions = "-Z"
+
+            if (noUnstableOptions !in libtestArgs) {
+                libtestArgs.add(noUnstableOptions)
+                libtestArgs.add("unstable-options")
+            }
+
+            if (noFormatJson !in libtestArgs) libtestArgs.add(noFormatJson)
```

Comment:
how does libtest react to multiple `--format` options? Like `cargo test -- --format=json --format=something`